# Bangladesh Housing & Population Census - Exploratory Data Analysis


## 1. Setup & Data Loading

In [1]:
import pandas as pd

import numpy as np


df=pd.read_csv("housing_census.csv")

## Checking for missing values in columns and the entire dataset.

In [2]:
print("Missing Values in each column:")

print(df.isnull().sum())



Missing Values in each column:
Division                                                   0
District                                                   0
Division_Geocode                                           0
District_Geocode                                           0
Household_Total                                            0
                                                          ..
Number of Person & Avg HH Size_# of HH with 8-Person       0
Number of Person & Avg HH Size_# of HH with 9-Person       0
Number of Person & Avg HH Size_# of HH with 10-Person      0
Number of Person & Avg HH Size_# of HH with 10-Person +    0
Number of Person & Avg HH Size_Average Household Size      0
Length: 445, dtype: int64


 There are not any column with missing values.

## Checking Duplicate rows

In [3]:
print("Number of duplicate rows:",df.duplicated().sum())



Number of duplicate rows: 0


## Printing out the  Column names

In [4]:
for i, column in enumerate(df.columns, start=1):
    print(f"{i}.{column}")

1.Division
2.District
3.Division_Geocode
4.District_Geocode
5.Household_Total
6.Population_Total
7.Household_Excluding_Slum_Floating
8.Population_Excluding_Slum_Floating
9.Household_Slum
10.Population_Slum
11.Household_Floating
12.Population_Floating
13.Population Density
14.Type of Household_Household_total_except_floating_#
15.Type of Household_Population__except_floating_#
16.Type of Household_Household_General_#
17.Type of Household_Population_General_#
18.Type of Household_Household_Institutional
19.Type of Household_Population_Institutional_#
20.Type of Household_Household_Other_#
21.Type of Household_Population_Other_#
22.Type of Dwelling, Sex & Dist_Population_Total_#
23.Type of Dwelling, Sex & Dist_Population_Male_#
24.Type of Dwelling, Sex & Dist_Population_Female_#
25.Type of Dwelling, Sex & Dist_Population_Hijra_#
26.Type of Dwelling, Sex & Dist_Population_Excluding_Slum_Floating_Total_#
27.Type of Dwelling, Sex & Dist_Population_Excluding_Slum_Floating_Male_#
28.Type of Dw

## Some features of the dataset

In [5]:
df.head(5)


,Division,District,Division_Geocode,District_Geocode,Household_Total,Population_Total,Household_Excluding_Slum_Floating,Population_Excluding_Slum_Floating,Household_Slum,Population_Slum,...,Number of Person & Avg HH Size_# of HH with 3-Person,Number of Person & Avg HH Size_# of HH with 4-Person,Number of Person & Avg HH Size_# of HH with 5-Person,Number of Person & Avg HH Size_# of HH with 6-Person,Number of Person & Avg HH Size_# of HH with 7-Person,Number of Person & Avg HH Size_# of HH with 8-Person,Number of Person & Avg HH Size_# of HH with 9-Person,Number of Person & Avg HH Size_# of HH with 10-Person,Number of Person & Avg HH Size_# of HH with 10-Person +,Number of Person & Avg HH Size_Average Household Size
0,Barishal,Barguna,10,4,255390,1010531,254895,1008695,451,1789,...,57097,68828,43330,20901,8395,3499,1583,772,767,3.92
1,Khulna,Bagerhat,40,1,408840,1613076,406058,1603126,2605,9763,...,96057,109638,68442,32081,12933,5722,2560,1324,1236,3.89
2,Barishal,Barishal,10,6,629626,2570446,620735,2537292,8160,31338,...,135271,165285,110132,54039,23613,10535,5289,2819,2981,4.02
3,Barishal,Bhola,10,9,449057,1932518,448944,1932138,74,339,...,89014,118802,92260,46329,20396,9347,4588,2375,2574,4.27
4,Barishal,Jhalokati,10,42,162401,661160,162212,660520,143,588,...,34540,41357,27852,14052,6339,3084,1529,815,904,4.02


In [6]:
df.tail(5)

,Division,District,Division_Geocode,District_Geocode,Household_Total,Population_Total,Household_Excluding_Slum_Floating,Population_Excluding_Slum_Floating,Household_Slum,Population_Slum,...,Number of Person & Avg HH Size_# of HH with 3-Person,Number of Person & Avg HH Size_# of HH with 4-Person,Number of Person & Avg HH Size_# of HH with 5-Person,Number of Person & Avg HH Size_# of HH with 6-Person,Number of Person & Avg HH Size_# of HH with 7-Person,Number of Person & Avg HH Size_# of HH with 8-Person,Number of Person & Avg HH Size_# of HH with 9-Person,Number of Person & Avg HH Size_# of HH with 10-Person,Number of Person & Avg HH Size_# of HH with 10-Person +,Number of Person & Avg HH Size_Average Household Size
59,Rangpur,Panchagarh,55,77,281627,1179843,280851,1176974,735,2827,...,56319,81106,53595,26636,11085,4651,2404,1281,1416,4.14
60,Rangpur,Rangpur,55,85,834307,3169614,826692,3141432,7447,28003,...,203689,231968,124275,52856,20364,8494,4142,2101,2248,3.72
61,Rangpur,Thakurgaon,55,94,382400,1533895,381262,1529809,1021,3930,...,82779,112589,66032,30798,11940,4898,2590,1297,1416,3.96
62,Sylhet,Sunamganj,60,90,528550,2695496,527545,2691484,865,3868,...,70460,104937,105541,78260,48226,27163,14596,8854,13911,5.09
63,Sylhet,Sylhet,60,91,746854,3857123,722553,3753617,23853,103032,...,102825,152780,143801,101254,62353,36293,21359,13723,24742,5.13


In [7]:
rows,cols=df.shape
print(f"Dataset Dimensions: {rows} rows * {cols} columns")

Dataset Dimensions: 64 rows * 445 columns


 Dataset Summary

Rows:     65 (64 administrative districts + 1 national aggregate)
Columns:  445 (housing metrics, population counts, and demographic indicators)
Format :  Wide-tabular datas

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 64 entries, 0 to 63
Columns: 445 entries, Division to Number of Person & Avg HH Size_Average Household Size
dtypes: float64(208), int64(234), str(3)
memory usage: 223.9 KB


In [9]:
df.describe()

,Division_Geocode,District_Geocode,Household_Total,Population_Total,Household_Excluding_Slum_Floating,Population_Excluding_Slum_Floating,Household_Slum,Population_Slum,Household_Floating,Population_Floating,...,Number of Person & Avg HH Size_# of HH with 3-Person,Number of Person & Avg HH Size_# of HH with 4-Person,Number of Person & Avg HH Size_# of HH with 5-Person,Number of Person & Avg HH Size_# of HH with 6-Person,Number of Person & Avg HH Size_# of HH with 7-Person,Number of Person & Avg HH Size_# of HH with 8-Person,Number of Person & Avg HH Size_# of HH with 9-Person,Number of Person & Avg HH Size_# of HH with 10-Person,Number of Person & Avg HH Size_# of HH with 10-Person +,Number of Person & Avg HH Size_Average Household Size
count,64.000000,64.000000,6.400000e+01,6.400000e+01,6.400000e+01,6.400000e+01,64.000000,64.000000,64.000000,64.000000,...,64.000000,64.000000,64.000000,64.000000,64.000000,64.000000,64.000000,64.000000,64.000000,64.000000
mean,36.406250,51.718750,6.407534e+05,2.580603e+06,6.327701e+05,2.553127e+06,7684.296875,27129.718750,298.968750,346.640625,...,140580.968750,163438.796875,104034.093750,51471.812500,23050.875000,10938.828125,5659.546875,3156.812500,4164.250000,4.031406
std,15.209353,27.003509,5.635301e+05,2.134836e+06,5.382841e+05,2.052937e+06,27837.256493,93746.357796,770.804115,853.589477,...,127738.850564,136359.196436,78102.420883,37512.175318,18268.022438,9525.834931,5253.640271,3221.510298,5208.502728,0.355335
min,10.000000,1.000000,1.061550e+05,4.811060e+05,1.060650e+05,4.810160e+05,0.000000,0.000000,0.000000,0.000000,...,19346.000000,24439.000000,19101.000000,11631.000000,4149.000000,1557.000000,688.000000,342.000000,252.000000,3.290000
25%,20.000000,31.500000,3.381995e+05,1.395069e+06,3.357908e+05,1.387069e+06,317.750000,1201.500000,69.250000,88.250000,...,72684.500000,92999.750000,61767.750000,29641.250000,12291.750000,5211.750000,2527.250000,1278.500000,1406.750000,3.805000
50%,40.000000,53.000000,5.225860e+05,2.156286e+06,5.212175e+05,2.148722e+06,1461.000000,5622.000000,127.500000,150.000000,...,102796.500000,134460.000000,92195.000000,43650.000000,17348.000000,8212.500000,3894.000000,1940.500000,2267.000000,3.985000
75%,50.000000,75.250000,7.503785e+05,2.955293e+06,7.452630e+05,2.935862e+06,3130.000000,12305.750000,275.750000,302.250000,...,165187.250000,184179.750000,123470.750000,58406.250000,26124.250000,11893.750000,6062.500000,3244.750000,3958.750000,4.150000
max,60.000000,94.000000,4.035241e+06,1.473470e+07,3.822371e+06,1.404069e+07,206900.000000,687534.000000,5970.000000,6477.000000,...,914430.000000,969773.000000,504971.000000,211944.000000,89097.000000,44943.000000,24934.000000,15625.000000,25295.000000,5.130000


## Unique values across all columns.

In [10]:
print(df.nunique(dropna=False).head(10))

Division                               8
District                              64
Division_Geocode                       8
District_Geocode                      64
Household_Total                       64
Population_Total                      64
Household_Excluding_Slum_Floating     64
Population_Excluding_Slum_Floating    64
Household_Slum                        63
Population_Slum                       63
dtype: int64


In [11]:
print(df.nunique(dropna=False).tail(10))

Number of Person & Avg HH Size_# of HH with 3-Person       64
Number of Person & Avg HH Size_# of HH with 4-Person       64
Number of Person & Avg HH Size_# of HH with 5-Person       64
Number of Person & Avg HH Size_# of HH with 6-Person       64
Number of Person & Avg HH Size_# of HH with 7-Person       64
Number of Person & Avg HH Size_# of HH with 8-Person       64
Number of Person & Avg HH Size_# of HH with 9-Person       64
Number of Person & Avg HH Size_# of HH with 10-Person      64
Number of Person & Avg HH Size_# of HH with 10-Person +    63
Number of Person & Avg HH Size_Average Household Size      51
dtype: int64


## Checking for inconsistencies in the dataset.

## Counting text columns

In [12]:
text_columns = df.select_dtypes(include=["object", "str"]).columns

print("Number of text columns:", len(text_columns))

Number of text columns: 3


## Lengths of the Columns and Their Data Types 

In [13]:
for column in text_columns:
    print(f"\n--- {column} ---")
    print(df[column].unique())


--- Division ---
<ArrowStringArray>
[  'Barishal',     'Khulna', 'Chattogram',   'Rajshahi',      'Dhaka',
    'Rangpur',     'Sylhet', 'Mymensingh']
Length: 8, dtype: str

--- District ---
<ArrowStringArray>
[        'Barguna',        'Bagerhat',        'Barishal',           'Bhola',
       'Jhalokati',       'Bandarban',      'Patuakhali',        'Pirojpur',
    'Brahmanbaria',        'Chandpur',          'Bogura',      'Chattogram',
         'Cumilla',     'Cox's Bazar',            'Feni',    'Khagrachhari',
      'Lakshmipur',       'Chuadanga',        'Noakhali',       'Rangamati',
           'Dhaka',        'Faridpur',         'Gazipur',       'Gopalganj',
     'Kishoreganj',       'Madaripur',        'Dinajpur',       'Manikganj',
      'Munshiganj', 'Chapainababganj',     'Narayanganj',       'Narsingdi',
         'Rajbari',      'Shariatpur',       'Gaibandha',         'Tangail',
         'Jashore',       'Jhenaidah',        'Habiganj',          'Khulna',
         'Kushtia', 

 ### Values that cannot be converted to numbers

In [14]:
column = "Household by Roof Material_Wood/Bamboo/Bichali/Golpata Palm Leaf/Polythen"


non_numeric = pd.to_numeric(df[column], errors="coerce").isna()

print(df.loc[non_numeric, column])

30    s
Name: Household by Roof Material_Wood/Bamboo/Bichali/Golpata Palm Leaf/Polythen, dtype: str


In [15]:
print(df.loc[non_numeric, ["Division", "District", column]])

   Division     District  \
30    Dhaka  Narayanganj   

   Household by Roof Material_Wood/Bamboo/Bichali/Golpata Palm Leaf/Polythen  
30                                                  s                         


In [16]:
print(df.loc[27:33, ["Division", "District", column]])

    Division         District  \
27     Dhaka        Manikganj   
28     Dhaka       Munshiganj   
29  Rajshahi  Chapainababganj   
30     Dhaka      Narayanganj   
31     Dhaka        Narsingdi   
32     Dhaka          Rajbari   
33     Dhaka       Shariatpur   

   Household by Roof Material_Wood/Bamboo/Bichali/Golpata Palm Leaf/Polythen  
27                                               0.26                         
28                                               1.23                         
29                                               0.43                         
30                                                  s                         
31                                               0.51                         
32                                               0.68                         
33                                               0.95                         


Non-Numeric String Error: Narayanganj District

## Figuring out the total number of numeric columns

In [17]:

numeric_columns = df.select_dtypes(include=["number"]).columns

print("Number of numeric columns:", len(numeric_columns))

print(df[numeric_columns].describe().T)

Number of numeric columns: 442
                                                    count          mean  \
Division_Geocode                                     64.0  3.640625e+01   
District_Geocode                                     64.0  5.171875e+01   
Household_Total                                      64.0  6.407534e+05   
Population_Total                                     64.0  2.580603e+06   
Household_Excluding_Slum_Floating                    64.0  6.327701e+05   
...                                                   ...           ...   
Number of Person & Avg HH Size_# of HH with 8-P...   64.0  1.093883e+04   
Number of Person & Avg HH Size_# of HH with 9-P...   64.0  5.659547e+03   
Number of Person & Avg HH Size_# of HH with 10-...   64.0  3.156812e+03   
Number of Person & Avg HH Size_# of HH with 10-...   64.0  4.164250e+03   
Number of Person & Avg HH Size_Average Househol...   64.0  4.031406e+00   

                                                             std    

## Checking if there are any negative values. 

In [18]:

numeric_columns = df.select_dtypes(include=["number"]).columns

negative_values = (df[numeric_columns] < 0).sum()

print(negative_values[negative_values > 0])

Population Growth Rate, Sex Rat_Overall_Avg_Population Growth     1
Population Growth Rate, Sex Rat_Rural_Avg. Population Growth     25
Population Growth Rate, Sex Rat_Urban_Avg__Population Growth      1
dtype: int64


## Checking if there are any infinite values. 

In [19]:

infinite_values = np.isinf(df[numeric_columns]).sum()

print(infinite_values[infinite_values > 0])

Series([], dtype: int64)


## Checking column with percentage and their validity

In [20]:
percentage_columns = [ column for column in df.columns if "%" in column or "percent" in column.lower() or "percentage" in column.lower()] 

print("Number of percentage columns:", len(percentage_columns))

for column in percentage_columns:
 print(column)

Number of percentage columns: 158
Type of Household, Dist and Loca_% Population_General HH
Type of Household, Dist and Loca_% Population_Institutional HH
Type of Household, Dist and Loca_% Population_other HH
Type of Household, Loca_Rural_% Population_General HH_Rural
Type of Household, Loca_Rural_% Population_Institutional HH_Rural
Type of Household, Loca_Rural_% Population_other HH_Rural
Type of Household, Loca_Urban.% Population_General HH_Urban
Type of Household, Loca_Urban.% Population_Institutional HH_Urban
Type of Household, Loca_Urban.% Population_other HH_Urban
15-24 Years NEET_%_Overall
15-24 Years NEET_%_male_Overall
15-24 Years NEET_%_female_Overall
15-24 Years NEET_%_rural_Overall
15-24 Years NEET_%_male_rural_Overall
15-24 Years NEET_%_female_rural_Overall
15-24 Years NEET_%_urban_Overall
15-24 Years NEET_%_male_urban_Overall
15-24 Years NEET_%_female_urban_Overall
15-24 Years NEET_%_15_19 year_overall
15-24 Years NEET_%_male_15_19 year
15-24 Years NEET_%_female_15_19 yea

In [21]:
invalid_percentage_columns = []

for column in percentage_columns:
    invalid = df[(df[column] < 0) | (df[column] > 100)]

    if not invalid.empty:
        invalid_percentage_columns.append(column)
        print(f"\nInvalid values in: {column}")
        print(invalid[["Division", "District", column]])

if not invalid_percentage_columns:
    print("All percentage columns contain valid values between 0% and 100%.")

All percentage columns contain valid values between 0% and 100%.


Here are 158 columns with percentages that contain valid values between 0% and 100%."